rfcs/modal-component.md

```diff
@@ -0,0 +1,154 @@
+* `onClose: function` - Optional
+  A callback that is invoked when the modal will close
+* `size: SIZE.{small|medium|large|full|cover|auto}|string|number`, Defaults to SIZE.default
+  Controls the size of the modal (primarily width). Can be a SIZE constant or css `width` property value.
+* `aria-labelledby: string` - Optional, Default to ''
+  If specified, will be passed to the dialog element
+* `aria-describedby: string` - Optional, Defaults to ''
+  If specified, will be passed to the dialog element
+* `aria-label: string` - Optional, Defaults to ''
+  If specified, will be passed to the dialog element (described the modal for screen readers)
+* `id: string` - Optional, Defaults to ''
+  If specified, will be passed to the dialog element
+* `role: 'dialog'|'alertdialog'|string` - Optional, Defaults to 'dialog'
+  Which accessibility role this modal should have
+* `mountNode: HTMLElement` - Optional, Defaults to document.body
+  Where to mount the modal
+* `overrides: {Root, Backdrop, Dialog, DialogInner, Close}` - Optional
+  Overrides for presentational components. See "Presentational Components Props API" below.
+  * `[ComponentName]: ReactComponent | {props: {}, style: {}, component: ReactComponent}` - Optional
+
+### <ModalHeader/> API
+
+TBD
+
+### <ModalBody/> API
+
+TBD
+
+### <ModalFooter/> API
+
+TBD
+
+### SIZE Constant
+
+* `small` - 400px width
+* `medium` - 550px width
+* `large` - 800px width
+* `full` - Span full viewport, minus some margin
+* `cover` - Span full viewport, no margin
+* `auto` - Based on content width
+
+^ Open to ideas here. Too many options?
+
+### Presentational Components Props API
+
+Next properties are passed to every presentational (styled) component
+
+* `$isOpen: boolean`
+* `$size: string|number`
+* `$role: string`
+* `$closable: boolean`
+* `$theme: theme`
+
+More props likely TBA as development continues
+
+### Usage
+
+Basic usage:
+
+```javascript
+import * as React from 'react';
+import {Modal, SIZE} from 'baseui/modal';
+
+class App extends React.Component {
+  state = {
+    open: true
+  };
+  render() {
+    return (
+      <div>
+        <Button onClick={() => this.setState({open: true})}>Open</Button>
+        <Modal
+          isOpen={this.state.open}
+          close={() => this.setState({open: false})}
+        >
+          {({close}) => (
+            <>
+              <ModalTitle>Whatsup!</ModalTitle>
+              <ModalBody>This is a modal</ModalBody>
+              <ModalFooter>
+                <Button type=”tertiary” onClick={close}>Close</Button>
+                <Button type=”primary” onClick={this.submit}>Let’s Go</Button>
+              </ModalFooter>
+            </>
+          )}
+        </Modal>
+      </div>
+
+    );
+  }
+}
+```
+
+### Key Design Decisions
+
+* Modal content will not render server side. This is because portals don't exist server-side, and realistically your page shouldn't be loading with a modal open.
```

Comment:
Wouldn't there be a case when you actually want to do it?
You get some user state from the server, if something matches some condition, open the modal on that page?
Or like the "What's New" modal?

Reply:
I suppose there may be valid cases for it. The question is how do you implement this safely? Normally you want to add modals as a direct child of (or close to) document.body to avoid [z-index conflicts with other elements on the page](https://codepen.io/anon/pen/qyGZqN), but you can't use portals server side. And if you change where it's mounted on the client, you get the almighty `React attempted to reuse markup in a container but the checksum was invalid` error.

Open to ideas if you have any.

Reply:
Eh, not sure.
One thing come to my mind is to use `componentDidMount` ?

```
constructor(props) {
  super(props);
  this.state = {
    isMounted: false,
  };
}

componentDidMount() {
  this.setState({isMounted: true});
}

render() {
  return this.state.isMounted && <SomePortalElement />;
}
```

Not sure whether it will give you that almighty error?

Reply:
>If portals disabled server-side by React for some reason, I think we should try to do it via DOM itself, although it's outside of React world a little.

It's not that "portals" are disabled. It's just that no document.body exists on the server since it's just rendering the root react component to a string. Can you clarify what you mean by "we should try to do it via DOM itself"?

Reply:
I thought you just cant use Portal at all there. So it's either goes to stringified version on server or will be mounted later via client-side scripts. Now I am confused and try to identify problem here.
By DOM I meant something like 
```js
ReactDOM.render(modal, document.body);
```

Reply:
Yeah I guess you can't use portal there at all because it requires a dom node to mount into.

Code like the following is what causes the server/client mismatch error:
```
render() {
  if (__IS_BROWSER__) {
    return ReactDOM.createPortal(this.renderModal(), document.body);
  }
  return this.renderModal();
}
```

Ardo's example code is likely the approach we should take, however since componentDidMount isn't called on the server, the modal still will not render on the server.

I don't love the idea of rendering as a direct child on the server, and then updating to a portal on the client, it feels like there could be z-index conflicts or jank before/after the switch.

My vote is that we just support client-rendering for now, and then when a customer requests support for server-side rendering, we add extra props/styles where necessary to facilitate this.

Reply:
Ok, so the issue is that in Portals API there is no "stringify and append to <body> tag" method for serverside rendering. So Ardo's will work, but it's almost the same as to say "if __BROWSER__". Can we try to apply decorator here only for serverside? So it will silently amend logic and modal code itself will not be affected?

https://medium.com/@gigobyte/enhancing-react-components-with-decorators-441320e8606a

Reply:
@schnerd just found this:
https://github.com/mui-org/material-ui/blob/master/packages/material-ui/src/NoSsr/NoSsr.js

Similar concept, that is display it on componentDidMount

Reply:
Nice, that looks like a good approach.